language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/util/RegexpCompletionProvider.java

```diff
@@ -0,0 +1,688 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.ballerinalang.langserver.completions.providers.context.util;
+
+import io.ballerina.compiler.syntax.tree.NonTerminalNode;
+import io.ballerina.compiler.syntax.tree.ReBracedQuantifierNode;
+import io.ballerina.compiler.syntax.tree.ReFlagExpressionNode;
+import io.ballerina.compiler.syntax.tree.ReFlagsOnOffNode;
+import io.ballerina.compiler.syntax.tree.ReQuantifierNode;
+import io.ballerina.compiler.syntax.tree.ReUnicodeGeneralCategoryNode;
+import io.ballerina.compiler.syntax.tree.ReUnicodePropertyEscapeNode;
+import io.ballerina.compiler.syntax.tree.SyntaxKind;
+import io.ballerina.compiler.syntax.tree.TemplateExpressionNode;
+import org.ballerinalang.langserver.commons.BallerinaCompletionContext;
+import org.ballerinalang.langserver.commons.completion.LSCompletionItem;
+import org.ballerinalang.langserver.completions.SnippetCompletionItem;
+import org.ballerinalang.langserver.completions.providers.context.TemplateExpressionNodeContext.RegexTemplateNodeFinder;
+import org.ballerinalang.langserver.completions.util.ItemResolverConstants;
+import org.ballerinalang.langserver.completions.util.Snippet;
+import org.ballerinalang.langserver.completions.util.SnippetBlock;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+
+/**
+ * Util to generate completions for regexp.
+ *
+ * @since 2201.5.0
+ */
+public class RegexpCompletionProvider {
+
+    private static final List<String> wordSeparatorArray = Arrays.asList("`", "~", "!", "@", "#", "$", "%", "^", "&",
+            "*", "(", ")", "-", "=", "+", "[", "{", "]", "}", "\\", "|", ";", ":", "'", "\"", ",", ".", "<", ">", "/",
+            "?");
+
+    private static final HashSet<String> reFlags = new HashSet<>(Arrays.asList("i", "m", "s", "x"));
+
+    public static List<LSCompletionItem> getRegexCompletions(NonTerminalNode nodeAtCursor,
+                                                             BallerinaCompletionContext ctx,
+                                                             RegexTemplateNodeFinder nodeFinder) {
+        List<LSCompletionItem> completionItems = new ArrayList<>();
+        if (nodeAtCursor.kind() == SyntaxKind.LIST) {
+            nodeAtCursor.parent().accept(nodeFinder);
+        } else {
+            nodeAtCursor.accept(nodeFinder);
+        }
+        String resolvedWord = null;
+
+        if (nodeAtCursor.kind() == SyntaxKind.RE_QUOTE_ESCAPE
+                && nodeAtCursor.toSourceCode().equals(SyntaxKind.BACK_SLASH_TOKEN.stringValue())) {
```

Comment:
Can we avoid a string comparison and use `Token` at cursor and it's kind to do this check?

Reply:
No we cannot.